// @progress/kendo-react-dateinputs/dist/npm/calendar/utils/ScrollSync.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @hidden
 */
var ScrollSync = /** @class */ (function () {
    function ScrollSync(detailsMonthHeight, navigationMonthHeight, scrollDetails, scrollNavigation) {
        this.scrollDetails = scrollDetails;
        this.scrollNavigation = scrollNavigation;
        this.magnitude = Math.max(detailsMonthHeight / navigationMonthHeight, 1);
    }
    ScrollSync.prototype.syncDetails = function (navigationScrollTop) {
        if (this.isNavigationSynced) {
            this.isNavigationSynced = false;
            return;
        }
        this.scrollDetails(navigationScrollTop * this.magnitude);
        this.areDetailsSynced = true;
    };
    ScrollSync.prototype.syncNavigation = function (detailsScrollTop) {
        if (this.areDetailsSynced) {
            this.areDetailsSynced = false;
            return;
        }
        this.scrollNavigation(Math.floor(detailsScrollTop / this.magnitude));
        this.isNavigationSynced = true;
    };
    return ScrollSync;
}());
exports.default = ScrollSync;
